baseapp/abci.go

```diff
@@ -45,13 +45,15 @@ func (app *BaseApp) InitChain(req *abci.RequestInitChain) (*abci.ResponseInitCha
 	// On a new chain, we consider the init chain block height as 0, even though
 	// req.InitialHeight is 1 by default.
 	initHeader := cmtproto.Header{ChainID: req.ChainId, Time: req.Time}
-	app.initialHeight = req.InitialHeight
 
 	app.logger.Info("InitChain", "initialHeight", req.InitialHeight, "chainID", req.ChainId)
 
 	// Set the initial height, which will be used to determine if we are proposing
 	// or processing the first block or not.
 	app.initialHeight = req.InitialHeight
+	if app.initialHeight == 0 { // If initial height is 0, set it to 1
+		app.initialHeight = 1
+	}
```

Comment:
I would put this in the `BaseApp` constructor, not here. 

Also, can you describe what issue this was causing?

Reply:
I don't think we can do it in the constructor, we have to get it in InitChain which contains InitialHeight (the line above `app.initialHeight = req.InitialHeight`).

The issue here is the following:

1. InitChain is called with InitialHeight of 0 (according to CometBFT it 0 == 1 in this context) and `app.initialHeight` is set to 0.
2. After that `ProcessProposal` gets called with Height 1 (calling it with height 0 is invalid; we have that check in place)
3. `getContextForProposal` is called with height 1. Then height != app.initialHeight (1 != 0), so it doesn't return the right context (the one that contains all the state written during InitChain).

Writing this makes me doubt if PrepareProposal/ProcessProposal can actually get called with height 0, AFAIK no, but I'm checking Comet's code. Do you know this @alexanderbez ?

[See getContextForProposal ](https://github.com/cosmos/cosmos-sdk/blob/88f2c830b0ca3b5cb874adfaa9ad8234bf00453c/baseapp/abci.go#L1004)

Reply:
> PrepareProposal/ProcessProposal can actually get called with height 0

This should never be the case I'd imagine, but either way, yeah, let's keep it here.